retracing in strict doesn't like dataclass registration

avikchaudhuri opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #144487

Retracing in strict doesn't seem to like dataclass registration. Just refactoring some tests to make this explicit (whereas other export testing variants work fine).

Differential Revision: [D67985149](https://our.internmc.facebook.com/intern/diff/D67985149/)

avikchaudhuri committed:
retracing in strict doesn't like dataclass registration

Retracing in strict doesn't seem to like dataclass registration. Just refactoring some tests to make this explicit (whereas other export testing variants work fine).

Differential Revision: [D67985149](https://our.internmc.facebook.com/intern/diff/D67985149/)

[ghstack-poisoned]

avikchaudhuri committed:
Update on "retracing in strict doesn't like dataclass registration"


Retracing in strict doesn't seem to like dataclass registration. Just refactoring some tests to make this explicit (whereas other export testing variants work fine).

Differential Revision: [D67985149](https://our.internmc.facebook.com/intern/diff/D67985149/)

[ghstack-poisoned]

angelayi commented on test/export/test_export.py:
```diff
@@ -4369,8 +4380,36 @@ def forward(self, x):
             ):
                 self.assertTrue("source_fn_stack" in node.meta)
 
+    @testing.expectedFailureRetraceability  # T186979579
```
it seems like it's not just dataclasses -- also custom classes fail too

avikchaudhuri replied:
Oh, good to know.

avikchaudhuri replied:
How do you know this btw?

angelayi replied:
I tried it out -- because dataclass pytree registration just treats dataclasses as like any custom class, so I wanted to see if those would fail too. Turns out it does.

```python
class DataClass:
    a: torch.Tensor
    b: torch.Tensor

    def __init__(self, a, b):
        self.a = a 
        self.b = b

from torch.utils._pytree import register_pytree_node, SequenceKey

register_pytree_node(
    DataClass,
    lambda x: ([x.a, x.b], None),
    lambda x, c: DataClass(a=x[0], b=x[1]),
    serialized_type_name="test_export_api_with_dynamic_shapes.DataClass",
    flatten_with_keys_fn=lambda x: ([(SequenceKey(0), x.a), (SequenceKey(1), x.b)], [x.a, x.b])
)

class Foo(torch.nn.Module):
    def forward(self, inputs):
        return torch.matmul(inputs.a, inputs.b)

foo = Foo()
inputs = (DataClass(a=torch.randn(10, 2, 3), b=torch.randn(10, 3, 4)),)
batch = torch.export.Dim("batch")
efoo = torch.export.export(
    foo,
    inputs,
)

torch.export.export(efoo.module(), inputs)
```

angelayi replied:
Also after some investigation I think the issue is because `ep.module()` uses `fx_pytree.tree_flatten_spec`, and the custom classes here do not have a `flatten_fn_spec` registration. This is kind of a pain because users need to do another pytree registration for this fx_pytree. One possible fix is that we stop using `fx_pytree.tree_flatten_spec` in the result of `ep.module()`.

angelayi commented on test/export/test_export.py:
```diff
@@ -169,12 +168,24 @@ def foo_unbacked(x):
+@dataclass
+class Inp3:
+    f: torch.Tensor
+    p: torch.Tensor
```
what's the difference between Inp2 and Inp3?

avikchaudhuri replied:
Nothing other than field names, they just happened to be defined locally which defeats serde, so pulled them up.

angelayi commented on test/export/test_export.py:
```diff
@@ -6208,6 +6205,7 @@ def forward(self, x: Input):
                     else:
                         self.assertIsInstance(s, int)
 
+    @testing.expectedFailureRetraceability  # T186979579
```
why is this failing now?

avikchaudhuri replied:
lolwut, this should have been on the test above! merge weirdness

avikchaudhuri committed:
Update on "retracing in strict doesn't like dataclass registration"


Retracing in strict doesn't seem to like dataclass registration. Just refactoring some tests to make this explicit (whereas other export testing variants work fine).

Differential Revision: [D67985149](https://our.internmc.facebook.com/intern/diff/D67985149/)

[ghstack-poisoned]

avikchaudhuri committed:
Update on "retracing in strict doesn't like dataclass registration"


Retracing in strict doesn't seem to like dataclass registration. Just refactoring some tests to make this explicit (whereas other export testing variants work fine).

Differential Revision: [D67985149](https://our.internmc.facebook.com/intern/diff/D67985149/)

[ghstack-poisoned]
